### PR TITLE
Task #196024 :  User count dashboard

### PR DESCRIPTION
**Pull request by poojakarma**
1. Setup the proxy network.
2. Setup the environment file.
3. Create one function in service to call API to get tenants
4. Create one function in  Service to call TENANTS USER COUNT  with respect to Channel id's
5. By passing different channel id on API getting different tenant count 
6. Add Count values  by API count value in json object
7. Show data on UI inside table 
8. Add Loader for response waiting time 
9. Add warning msg if API not respond 
10. Add i18next feature for language Translation 
11. Add Search filter for get organization Name on table
12. Added pagination feature on table
13. Added sorting on columns on table

### CODE REVIEW

File: projects/sb-user-management/src/lib/user-count.service.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { Injectable } from '@angular/core';
```
Comment (nidhikachhawa24) [inline projects/sb-user-management/src/lib/user-count.service.ts]: @poojakarma 
Please move these service files in the user-count folder

File: projects/sb-user-management/src/lib/user-count.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { TestBed } from '@angular/core/testing';
```
Comment (nidhikachhawa24) [inline projects/sb-user-management/src/lib/user-count.service.spec.ts]: @poojakarma 
Please move these service files in the user-count folder

File: projects/sb-user-management/src/lib/user-count/user-count.component.css [hunk 3]
```diff
@@ -0,0 +1,17 @@
+h1{
+    color: green;text-align: center !important;
```
Comment (nidhikachhawa24) [inline projects/sb-user-management/src/lib/user-count/user-count.component.css]: @poojakarma 
Please don't use **! important** as it is considered as bad practice.

File: projects/sb-user-management/src/lib/user-count/user-count.component.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+@Component({
+  selector: 'lib-user-count',
+  templateUrl: './user-count.component.html',
+  styleUrls: ['./user-count.component.css']
+})
+export class UserCountComponent implements OnInit {
+
+  constructor(private userCountService: UserCountService) { }
+
+  tenantDetail: TenantDetails[]=tenanats;
```
Comment (nidhikachhawa24) [inline projects/sb-user-management/src/lib/user-count/user-count.component.ts]: @poojakarma 
Please correct the spelling **tenanats**. You can either user **tenantList** as the Json is referring to the list of tenants

File: projects/sb-user-management/src/lib/user-count/user-count.component.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  ngOnInit(): void {
+    this.getTenantUserCount();
+  }
+  getTenantUserCount() {
+    this.tenantDetail.map((tenant) => {
+      this.userCountService.getUserCountOfaTenant(tenant.channelId).subscribe((counttenant)=> {
+        const data : any= counttenant;
```
Comment (nidhikachhawa24) [inline projects/sb-user-management/src/lib/user-count/user-count.component.ts]: @poojakarma 
Why you are using type **any** here?

File: proxy.conf.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+const PROXY_CONFIG = [
+    {
+        context: [
+            "/content-plugins",
```
Comment (gouravmore) [inline proxy.conf.js]: Need to remove this unwanted routes for proxy

File: src/app/sb-admin/components/dashboard/dashboard.component.html [hunk 7]
```diff
@@ -1,38 +1,29 @@
...
+                <ng-template pTemplate="caption">
+                    <span class="p-input-icon-left">
+                        <i class="pi pi-search"> </i>
+                        <input type="text" pInputText placeholder="Organization Name"
+                            (input)="myTab.filterGlobal($any($event.target).value,'contains')" />
+                    </span>
+                </ng-template>
+                <ng-template pTemplate="header">
+                    <tr>
+                        <th pSortableColumn="orgName">Organization Name<p-sortIcon field="orgName"></p-sortIcon></th>
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.html]: Need to use i18 lang constants and files for UI labels

File: src/app/sb-admin/components/dashboard/dashboard.component.html [hunk 8]
```diff
@@ -1,38 +1,29 @@
...
+                        <th pSortableColumn="userCount">User Registered<p-sortIcon field="userCount"></p-sortIcon></th>
+                    </tr>
+                </ng-template>
+                <ng-template pTemplate="body" let-user>
+                    <tr>
+                        <td style="width: 35%; min-width: 7rem;">{{user.orgName}}</td>
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.html]: remove the inline css from here and add to the component CSS file

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 9]
```diff
@@ -2,8 +2,16 @@ import { Component, OnInit, OnDestroy } from '@angular/core';
 import { MenuItem } from 'primeng/api';
 import { User } from '../../api/user';
 import { UserService } from '../../service/user.service';
-import { Subscription } from 'rxjs';
+import { map, Subscription } from 'rxjs';
 import { LayoutService } from 'src/app/layout/service/app.layout.service';
+import { UserCountService } from '../../service/user-count.service';
+
+
+interface TenantDetails {
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: can you move this interface to interface file

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 10]
```diff
@@ -20,25 +28,57 @@ export class DashboardComponent implements OnInit, OnDestroy {
...
-            { label: 'ORG_1', icon: 'bg-blue-100' },
-            { label: 'ORG_2', icon: 'bg-green-100' },
-            { label: 'ORG_3', icon: 'bg-cyan-100' },
-            { label: 'ORG_4', icon: 'bg-purple-100' },
-            { label: 'ORG_5', icon: 'bg-blue-100' },
-            { label: 'ORG_6', icon: 'bg-green-100' },
-            { label: 'ORG_7', icon: 'bg-cyan-100' }
+
+        this.cols = [
+            { field: 'orgName', header: 'Organization Name' },
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: Need to use the language constants here

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 11]
```diff
@@ -20,25 +28,57 @@ export class DashboardComponent implements OnInit, OnDestroy {
...
         ];
+
+        this.getTenant().subscribe((data: any) => {
+            if (data) {
+
+                this.getTenantUsercount(data);
+            }
+        });
+
+        //this.userService.getUsers().then(data => this.users = data);
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: if it is not required - remove it

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 12]
```diff
@@ -20,25 +28,57 @@ export class DashboardComponent implements OnInit, OnDestroy {
...
+    }
+
+    //Get all tenant data
+    getTenant() {
+        return this.userCountService.getTenant().pipe(
+            map((data: any) => {
+                this.tenantDetail = data.result.response.content;
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: What if the result does not have a response? Use Safe Navigation Operator here

File: src/app/sb-admin/service/user-count.service.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+  constructor(private http: HttpClient) { }
+
+  getTenant(): Observable<Object> {
+
+    let header = new HttpHeaders({
+      "Content-Type": 'application/json'
+    })
+
+    let body = {
```
Comment (gouravmore) [inline src/app/sb-admin/service/user-count.service.ts]: the body should be defined in component not in service file

File: src/app/sb-admin/service/user-count.service.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+          "isRootOrg"
+        ],
+        "sortBy": {
+          "createdDate": "Desc"
+        },
+        "limit": 1002
+      }
+
+    }
+    return this.http.post('api/org/v1/search', body, { headers: header })
```
Comment (gouravmore) [inline src/app/sb-admin/service/user-count.service.ts]: Use URL config file and remove hardcoded urls

File: src/app/sb-admin/service/user-count.service.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  getUserCountOfaTenant(channelId: string): Observable<Object> {
+    let header = new HttpHeaders({
+      "Content-Type": 'application/json',
+      "Authorization": environment.authKey,
+      "x-authenticated-user-token": environment.userToken,
+    })
+
+    let body = {
```
Comment (gouravmore) [inline src/app/sb-admin/service/user-count.service.ts]: the body should be defined in component not in service file

File: src/app/sb-admin/service/user-count.service.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+          "email",
+          "phone",
+          "createdDate",
+          "roles",
+          "managedBy"
+        ],
+        "limit": 10
+      }
+    }
+    return this.http.post('api/user/v1/search', body, { headers: header })
```
Comment (gouravmore) [inline src/app/sb-admin/service/user-count.service.ts]: Use URL config file and remove hardcoded URLs

**Review by gouravmore — CHANGES_REQUESTED** [review]

File: proxy.conf.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
+const PROXY_CONFIG = [
+    {
+        context: [
+            "/api",
+            "/assets",
+            "/search"
+        ],
+        "target": "https://sunbirdsaas.com",
```
Comment (manojLondhe) [inline proxy.conf.js]: get this URL from config

File: src/app/app-routing.module.ts [hunk 18]
```diff
@@ -19,5 +19,5 @@ import { AppLayoutComponent } from "./layout/app.layout.component";
     ],
     exports: [RouterModule]
 })
-export class AppRoutingModule {
+export class AppRoutingModule {
```
Comment (manojLondhe) [inline src/app/app-routing.module.ts]: remove trailing whitespace from all files

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 19]
```diff
@@ -1,46 +1,133 @@
...
+
+  ngOnInit() {
+
+    //i18n
+    $(document).ready(function () {
+      $("#eng").click(function () {
+        localStorage.setItem('lang', 'en');
+        document.location.reload();
+      });
+      $("#de").click(function () {
```
Comment (manojLondhe) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: what is #de ?

File: src/locales/en/en.json [hunk 20]
```diff
@@ -0,0 +1,6 @@
+{
+  "DASHBOARD":"Dashboard",
+  "DASHBOARD_ORGANIZATION_NAME": "Organization Name",
+  "DASHBOARD_USER_REGISTERED": "User Registered",
```
Comment (manojLondhe) [inline src/locales/en/en.json]: user or users?

File: src/locales/ta/ta.json [hunk 21]
```diff
@@ -0,0 +1,6 @@
+{
+  "DASHBOARD":"Dashboard",
+  "DASHBOARD_ORGANIZATION_NAME": "Organization Name",
+  "DASHBOARD_USER_REGISTERED": "User Registered",
```
Comment (manojLondhe) [inline src/locales/ta/ta.json]: user or users?

**Review by manojLondhe — CHANGES_REQUESTED** [review]

File: angular.json [hunk 22]
```diff
@@ -69,6 +69,10 @@
         },
         "serve": {
           "builder": "@angular-devkit/build-angular:dev-server",
+          "options": {
+            "proxyConfig": "proxy.conf.js"
+          },
+
```
Comment (gouravmore) [inline angular.json]: remove extra line

File: proxy.conf.js [hunk 23]
```diff
@@ -0,0 +1,15 @@
+const PROXY_CONFIG = [
+    {
+        context: [
+            "/api",
+            "/assets",
+            "/search"
+
```
Comment (gouravmore) [inline proxy.conf.js]: remove extra line

File: src/app/app.module.ts [hunk 24]
```diff
@@ -4,17 +4,66 @@ import { AppRoutingModule } from './app-routing.module';
 import { AppLayoutModule } from './layout/app.layout.module';
 import { NotfoundComponent } from './sb-admin/components/notfound/notfound.component';
 import { UserService } from './sb-admin/service/user.service';
+import { UserCountService } from './sb-admin/service/user-count.service';
+//i18next
```
Comment (gouravmore) [inline src/app/app.module.ts]: is this required?

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 25]
```diff
@@ -1,46 +1,126 @@
...
+import $ from 'jquery';
 
 @Component({
-    templateUrl: './dashboard.component.html',
+  templateUrl: './dashboard.component.html',
 })
 export class DashboardComponent implements OnInit, OnDestroy {
 
-    items!: MenuItem[];
+  subscription!: Subscription;
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: what is this subscription about?

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 26]
```diff
@@ -1,46 +1,126 @@
...
+  tenantUserCount: any[] = [];
 
-    constructor(private userService: UserService, public layoutService: LayoutService) {
-    }
+  constructor(private userService: UserService, public layoutService: LayoutService, private userCountService: UserCountService) 
+  {}
+
+  ngOnInit() {
+    //i18n
+    $(document).ready(function () {
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts:22]: Is this still required?

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 27]
```diff
@@ -1,46 +1,126 @@
...
+      });
+    });
+    $(document).ready(function () {
+      $("#flip").click(function () {
+        $("#panel").slideToggle("fast");
+      });
+    });
+
+    this.cols = [
+      { field: 'orgName', header: 'Organization Name' },
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: This should come from the config

File: src/app/sb-admin/components/dashboard/dashboard.module.ts [hunk 28]
```diff
@@ -9,6 +9,52 @@ import { ButtonModule } from 'primeng/button';
 import { StyleClassModule } from 'primeng/styleclass';
 import { PanelMenuModule } from 'primeng/panelmenu';
 import { DashboardsRoutingModule } from './dashboard-routing.module';
+import { InputTextModule } from 'primeng/inputtext';
+import { MessagesModule } from 'primeng/messages';
+import { MessageModule } from "primeng/message";
+//i18next
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.module.ts]: is this comment required?

File: src/locales/ta/ta.json [hunk 29]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (gouravmore) [inline src/locales/ta/ta.json]: You can remove this ta.json as we don't have any translation for this

File: projects/sb-user-management/src/lib/user-count.service.ts (hunk 1 above)
Comment (poojakarma) [reply projects/sb-user-management/src/lib/user-count.service.ts]: Resolved

File: projects/sb-user-management/src/lib/user-count.service.spec.ts (hunk 2 above)
Comment (poojakarma) [reply projects/sb-user-management/src/lib/user-count.service.spec.ts]: Resolved

File: projects/sb-user-management/src/lib/user-count/user-count.component.css (hunk 3 above)
Comment (poojakarma) [reply projects/sb-user-management/src/lib/user-count/user-count.component.css]: Removed

File: projects/sb-user-management/src/lib/user-count/user-count.component.ts (hunk 4 above)
Comment (poojakarma) [reply projects/sb-user-management/src/lib/user-count/user-count.component.ts]: Done

File: projects/sb-user-management/src/lib/user-count/user-count.component.ts (hunk 5 above)
Comment (poojakarma) [reply projects/sb-user-management/src/lib/user-count/user-count.component.ts]: for getting API response it can be any

File: proxy.conf.js (hunk 6 above)
Comment (poojakarma) [reply proxy.conf.js]: Removed

File: src/app/sb-admin/service/user-count.service.ts (hunk 16 above)
Comment (poojakarma) [reply src/app/sb-admin/service/user-count.service.ts]: Done

File: src/locales/en/en.json (hunk 20 above)
Comment (poojakarma) [reply src/locales/en/en.json]: changed

File: src/locales/ta/ta.json (hunk 21 above)
Comment (poojakarma) [reply src/locales/ta/ta.json]: changed

File: src/app/sb-admin/components/dashboard/dashboard.component.ts (hunk 26 above)
Comment (gouravmore) [reply src/app/sb-admin/components/dashboard/dashboard.component.ts:22]: @poojakarma Is this code required?

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 30]
```diff
@@ -1,46 +1,98 @@
...
+          "id",
+          "channel",
+          "orgName",
+          "externalId",
+          "isRootOrg"
+        ],
+        "sortBy": {
+          "createdDate": "Desc"
+        },
+        "limit": 1002
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: @nidhikachhawa24 @poojakarma can you check this limit parameter to get tenant API? For me looks likes it is hardcoded.

File: src/app/sb-admin/components/dashboard/dashboard.component.ts [hunk 31]
```diff
@@ -1,46 +1,98 @@
...
+            "lastName",
+            "userName",
+            "id",
+            "email",
+            "phone",
+            "createdDate",
+            "roles",
+            "managedBy"
+          ],
+          "limit": 10
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.component.ts]: @poojakarma What if we need more that 10 user count

File: src/app/sb-admin/components/dashboard/dashboard.module.ts [hunk 32]
```diff
@@ -9,6 +9,47 @@ import { ButtonModule } from 'primeng/button';
...
 import { PanelMenuModule } from 'primeng/panelmenu';
 import { DashboardsRoutingModule } from './dashboard-routing.module';
+import { InputTextModule } from 'primeng/inputtext';
+import { MessagesModule } from 'primeng/messages';
+import { MessageModule } from "primeng/message";
+import { APP_INITIALIZER, LOCALE_ID } from '@angular/core';
+import { I18NextModule, ITranslationService, I18NEXT_SERVICE, defaultInterpolationFormat } from 'angular-i18next';
+import en from "src/locales/en/en.json";
+
+export function appInit(i18next: ITranslationService) {
```
Comment (gouravmore) [inline src/app/sb-admin/components/dashboard/dashboard.module.ts:19]: Is this required here?

**Pull request merged**